src/Microsoft.Azure.WebJobs.Extensions.PostgreSql/PostgreSqlAsyncCollector.cs

```diff
@@ -34,92 +62,282 @@ internal class PostgreSqlAsyncCollector<T> : IAsyncCollector<T>, IDisposable
+        }
+
+        private NpgsqlCommand CreateBatchInsertCommand(IEnumerable<T> batch, BatchInsertCommandComponents components)
+        {
+
+            NpgsqlCommand command = new NpgsqlCommand();
+
+            if (batch.Count() == _batchSize)
+            {
+                command.CommandText = _fullBatchCommandText;
+            }
+            else
             {
-                throw new ArgumentException(connectionString == null ? $"ConnectionStringSetting '{connectionStringSetting}' is missing in your function app settings, please add the setting with a valid PostgreSQL connection string." :
-                $"ConnectionStringSetting '{connectionStringSetting}' is empty in your function app settings, please update the setting with a valid PostgreSQL connection string.");
+                command.CommandText = $"{components.InsertClause} {GenerateValuesClause(batch.Count())} {components.ConflictClause}";
             }
-            return connectionString;
+
+            //now add parameters
+            command.Parameters.AddRange(CreateParameters(batch));
+
+            return command;
         }
 
-        /// <summary>
-        /// Adds an item to this collector that is processed in a batch along with all other items added via
-        /// AddAsync when <see cref="FlushAsync"/> is called. Each item is interpreted as a row to be added to the SQL table
-        /// specified in the SQL Binding.
-        /// </summary>
-        /// <param name="item"> The item to add to the collector </param>
-        /// <param name="cancellationToken">The cancellationToken is not used in this method</param>
-        /// <returns> A CompletedTask if executed successfully </returns>
-        public async Task AddAsync(T item, CancellationToken cancellationToken = default)
+        private Array CreateParameters(IEnumerable<T> batch)
         {
-            //unimplemented
-            await Task.Delay(0);
+            List<NpgsqlParameter> parameters = new List<NpgsqlParameter>();
+            int i = 0;
+            foreach (T item in batch)
+            {
+                foreach (var property in _generic_properties)
+                {
+                    parameters.Add(new NpgsqlParameter($"@{property.Name}{i}", property.GetValue(item)));
+                }
+                i++;
+            }
+
+            return parameters.ToArray();
         }
 
+
         /// <summary>
         /// </summary>
         /// <param name="cancellationToken">The cancellationToken is not used in this method</param>
         /// <returns> A CompletedTask if executed successfully. If no rows were added, this is returned
         /// automatically. </returns>
         public async Task FlushAsync(CancellationToken cancellationToken = default)
         {
+            await _rowLock.WaitAsync(cancellationToken);
+            try
+            {
+                if (_rows.Count != 0)
+                {
+                    await SetColumnData();
```

Comment:
Is this method used to refresh table and column metadata before running the validity check? I was wondering all the rows are added by 'AddAsync' only to return exception during validity check if there's any changes in the schema. This is an open ended question and if you see any other way to ensure the validity beforehand, that would be great.